### ISSUE TITLE
Improve the message about warnings

### ISSUE THREAD

**Issue by ayberkt** [tectonic-typesetting/tectonic#577]
When warnings are issued during compilation by the TeX engine, `tectonic` prints the following message:

```
warning: warnings were issued by the TeX engine; use --print and/or --keep-logs for details.
```

It seems to me that `tectonic` should be printing out more information about the warning.
Seeing that `tectonic` can point out that warnings have occurred, it must be parsing or otherwise searching the output of the TeX engine. In this case, wouldn't it be quite easy to print out the line that makes `tectonic` think that a warning has occurred?

Of course, one can easily find the warning by using `--keep-logs` and then searching for "Warning" in the generated log but there are at least two reasons why this is not ideal:

1. It's much more convenient to not have to look at the log file.
2. Sometimes, `tectonic` tells me that warnings were issued but I cannot find anything involving a warning in the log file. Therefore I would like to understand what makes `tectonic` decide that a warning has been issued.

In fact, I have recently been compiling a document with `tectonic` during the compilation of which I keep getting the message about warnings despite the log not mentioning any warning. I made [a post about this problem](https://tectonic.newton.cx/t/tectonic-says-warnings-were-issued-but-i-cannot-find-those-warnings/231) in the forum.

**Comment by pkgw** [comment]
Agreed that the current state of affairs is less than ideal.

What drives the current behavior is that that the TeX engine exits with a status code indicating the final warning/error level ([tt_history_t](https://github.com/tectonic-typesetting/tectonic/blob/master/tectonic/core-bridge.h#L13) type). Unfortunately, there is nothing rigorous in the code that associates the setting of that status code with the printing of a specific message. So, to be able to *reliably* associate the warning status with a message, one would have to go through the C code and modify all of the paths that set the warning status to use the `ttstub_issue_warning` function at some point. No one's done that work as of yet.

From a quick grep through the sources, there are only a few places that seem to explicitly be setting the `history` variable to warning status, so maybe this would actually be easier than I realized!

Alternatively / additionally, one could imagine adding a layer that simply scanned the output of the TeX engine and printed lines including the text "warning", "error", etc. This might be helpful depending on how exactly LaTeX emits warning and error messages -- if the messages appear as generic print commands as far as the underlying TeX engine is concerned, then annotations in the C code aren't going to help much.

**Comment by ayberkt** [comment]
@pkgw thanks for the detailed explanation.

> Alternatively / additionally, one could imagine adding a layer that simply scanned the output of the TeX engine and printed lines including the text "warning", "error", etc. This might be helpful depending on how exactly LaTeX emits warning and error messages -- if the messages appear as generic print commands as far as the underlying TeX engine is concerned, then annotations in the C code aren't going to help much.

I find the TeX engine output quite confusing. For instance, are overfull/underfull "warnings" considered actual warnings? It seems like they should be, but the output does not use the word "warning" for them.

By the way, would you mind taking a look at [this post](https://tectonic.newton.cx/t/tectonic-says-warnings-were-issued-but-i-cannot-find-those-warnings/231) on the forums? It seems to me that this is a bug and if you can confirm that I will file an issue for it.

**Comment by pkgw** [comment]
@ayberkt Yes, the output is indeed confusing in many ways! One of my goals for Tectonic is to gradually supersede the "classic" TeX output with reports that are much more helpful.

I believe that overfull/underfull warnings count as "actual" warnings because the `hpack` function that reports the issues ([xref1]) calls `begin_diagnostic` ([xref2]) which is one of the places that sets the warning flag.

Apologies for being slow on the forums — my ability to spend time on Tectonic has unfortunately been very limited for a while.

[xref1]: https://github.com/tectonic-typesetting/tectonic/blob/master/tectonic/xetex-xetex0.c#L12129
[xref2]: https://github.com/tectonic-typesetting/tectonic/blob/master/tectonic/xetex-xetex0.c#L2094

**Comment by ayberkt** [comment]
@pkgw 

> One of my goals for Tectonic is to gradually supersede the "classic" TeX output with reports that are much more helpful.

That is a great idea and might become the second most compelling reason to use tectonic after the fact that it is "spelled and pronounced like a regular word because it is one" :smile:.

I would be happy to help as much as I can.

**Comment by pkgw** [comment]
@ayberkt All contributions are welcome, no matter how small! I haven't been able to put much time into the project lately but I try to at least merge pull requests.

**Comment by ralismark** [comment]
Hi, I'd be interesting in contributing and having a go at this issue, though I'm not that familiar with the codebase. Would this just involve adding `ttstub_issue_warning` calls when `history` gets set?

**Comment by pkgw** [comment]
@ralismark Thanks! Yes, I think that's the rough outline of what needs to happen. A good place to start might be the code path that I mentioned on my April 29 comment — of course under/overfull warnings are super common, and I believe that the `begin_diagnostic` function is used pretty widely, so if you figure out a good pattern for how to properly extract the warning there, that will be a very good start.

**Comment by pkgw** [comment]
@ralismark I guess I think it would probably be more helpful to have finer-grained issues for capturing specific categories of warnings/errors to capture? There are a *lot* of ways to trigger an error so this is probably going to be a long-term project.
